Support listing and exporting of recorded segments

jflevesque-genetec opened the pull request:
Add new APIs in the Search and Recording specification to be able to list and export recorded segments towards a cloud storage.

jflevesque-genetec committed:
Define new ListRecordingSegments API

jflevesque-genetec committed:
Add export segment API

jflevesque-genetec committed:
Update wsdl

jflevesque-genetec committed:
Update doc and add storage strategy

jflevesque-genetec committed:
Add annex

jflevesque-genetec commented on doc/RecordingControl.xml:
```diff
@@ -1595,6 +1595,85 @@ Change Request 2061, 2063, 2065, 2109</revremark>
+          <term>request</term>
+          <listitem>
+            <para role="param">Time [tt:DateTimeRange]</para>
+            <para role="param">JobToken [tt:RecordingJobReference]</para>
```
During the Krakow F2F, members have raised the question why we are not using RecordingConfiguration tokens instead. The issue we see is that without the JobToken, there is no way for the client to specify which Source (media profile) we want to list or export recorded segments for.

jflevesque-genetec replied:
Based on feedback received, updated the RecordingConfiguration to include the VideoSourceToken inside the RecordingConfiguration and use the RecordingToken instead.

jflevesque-genetec committed:
Change segments API to use RecordingToken instead of RecordingJobToken

jflevesque-genetec committed:
Add video source token to the RecordingSourceInformation

jflevesque-genetec committed:
Update search API to have paging for ListRecordedSegments

jflevesque-genetec committed:
Update ExportRecordedSegments API to have paging

jflevesque-genetec commented:
Updated the proposal based on feedback received since the F2F. Added "paging" mechanism to both the search and export operations.

HansBusch commented on doc/RecordingControl.xml:
```diff
@@ -1595,6 +1595,89 @@ Change Request 2061, 2063, 2065, 2109</revremark>
+            <para role="param">RecordingToken [tt:RecordingReference]</para>
+            <para role="param">Track [tt:TrackType]</para>
+            <para role="text">An optional field if specified indicates which track is to be exported.</para>
+            <para role="param">OmitSequenceNumber [xs:boolean]</para>
```
Purpose of OmitSequenceNumber unclear. If really needed the definition should be positive and default true.

jmelancongen replied:
It is worded negatively because it overrides what is the default behavior of the cloud upload to specify the "sequence number" of the file. Which now that I review, is actually named `counter` in the Recording Control spec.
So the parameter here would be named `OmitSegmentCounter` probably. And a clearer description.

jflevesque-genetec replied:
Renamed field to ForceZeroCounter and referenced the annex.

HansBusch commented on doc/RecordingSearch.xml:
```diff
@@ -1389,6 +1389,67 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
       </variablelist>
     </section>
+    <section>
+      <title>FindRecordedSegments</title>
```
Is this method needed for export functionality?

jmelancongen replied:
I generally comes first, when the clients want to know what is available nad in what sections, that it can then export more intelligently based on what it already has, and what users need
For our precise usecase, it also permits creation of an HLS playlist from those segments. I believe it also enables creating DASH manifests or other similar techs

HansBusch commented on doc/RecordingSearch.xml:
```diff
@@ -1486,6 +1547,107 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
       <para>The definition is available via the link &lt;<link xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="https://www.onvif.org/specs/srv/rsrch/ONVIF-RecordingSearch-Service-Spec-v1606.pdf">https://www.onvif.org/specs/srv/rsrch/ONVIF-RecordingSearch-Service-Spec-v1606.pdf</link>&gt;.</para>
     </section>
   </appendix>
+  <appendix xml:id="_refSegmentExport">
```
Shouldn't be this section be defined in Recording Control Service?

jmelancongen replied:
Half the calls are in search and the other half in control so we were a bit unclear on where would be the best place for it. Open to switching it

jflevesque-genetec replied:
Moved from Search to Recording Control

HansBusch commented on doc/RecordingControl.xml:
```diff
@@ -1595,6 +1595,89 @@ Change Request 2061, 2063, 2065, 2109</revremark>
+    <section>
+      <title>ExportRecordedSegments</title>
+      <para>Exports the selected recorded segments (from existing recorded data) to the storage attached to the given recording job.</para>
+      <para>The StoragePath shall be the path as defined in the Object Storage annex appended to the StorageUri from the StorageConfiguration.</para>
```
Propose to add an optional field Target so that this command also allows to export to other configured storage location.

jmelancongen replied:
This field would allow overriding the StorageConfiguration part of the configuration? I feel it couldn't be the whole Target section since that also contains the SegmentDuration parts, which if they change impacts of the files are split, and would break a device which decided to record directly in the right segment length I believe?

jmelancongen replied:
From F2F feedback: Add possibility to optionally override the StorageConfiguration during the Export call (no change to prefix/postfix and all other Target properties)

jflevesque-genetec replied:
Done

HansBusch reviewed (commented):
Started review with commenting document changes.

venki5685 commented on wsdl/ver10/schema/onvif.xsd:
```diff
@@ -7636,6 +7636,11 @@ and sample rate. </xs:documentation>
 			</xs:element>
+			<xs:element name="VideoSourceToken" type="tt:ReferenceToken" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>The video source token of all recordings that will use this configuration.</xs:documentation>
```
1. VideoSourceToken parameter in RecordingSourceInformation makes the recording container only record the data of this particular video source.  As Per ONVIF Recording Control specification, the recording container can support recording two video tracks, one containing a low resolution or low frame rate stream and one containing a high resolution or high frame rate stream of the same video source or different video source. 
2.  RecordingJobSource have Source Token i.e Profile token tells about which media should be recorded to the recording container. If the recording container already associated with video source via RecordingSourceInformation and if the video source in recording container is  different from video source present in RecordingJobSource Profile token, it creates in-computability issue and the job cannot record the particular media profile data to the specific recording container.

jflevesque-genetec replied:
Can you explain how 1. can be done with the current specification, because to my knowledge, you need to specify a MediaProfile which can only have 1 VideoEncoderConfiguration assigned to it. So you wouldn't be able to record different resolutions or video sources using only 1 RecordingJob.

2. Yes, this would be a new validation that would have to be done by the device to keep things in sync
3. Why wouldn't it be possible to change the SourceInformation?

venki5685 replied:
1. we can change media profile assigned to recording job, where the new media profile points to different video encoder, video source from the previous media profile.
3. After changing recording source information in the recording container, recorded data with old source in the container will reflect with new recording source when loading the recording timeline, that is not expected behavior.

jflevesque-genetec committed:
Add optional override storage configuration for export segments

jflevesque-genetec commented on wsdl/ver10/recording.wsdl:
```diff
@@ -10,8 +10,9 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
-		<xs:schema targetNamespace="http://www.onvif.org/ver10/recording/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" version="23.06">
+		<xs:schema targetNamespace="http://www.onvif.org/ver10/recording/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:tds="http://www.onvif.org/ver10/device/wsdl" xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" version="23.06">
 			<xs:import namespace="http://www.onvif.org/ver10/schema" schemaLocation="../ver10/schema/onvif.xsd"/>
+			<xs:import namespace="http://www.onvif.org/ver10/device/wsdl" schemaLocation="../ver10/device/wsdl/devicemgmt.wsdl"/>
```
@HansBusch By adding the override, this pulled the device management namespace into this WSDL. Do I need to import the namespace or specifying the namespace is sufficient?

jflevesque-genetec replied:
As discussed, removed dependency on DeviceMgmt WSDL and used storage token instead.

jflevesque-genetec committed:
Remove dependency on Device Mgmt by using StorageToken

jflevesque-genetec committed:
Apply F2F suggestions for export

jflevesque-genetec committed:
Apply suggestion for Search API from F2F

jflevesque-genetec committed:
Move ListRecordedSegments from RecordingSearch to RecordingControl

jflevesque-genetec committed:
Changed should to shall in export annex

venki5685 commented on wsdl/ver10/schema/onvif.xsd:
```diff
@@ -7723,6 +7728,11 @@ and sample rate. </xs:documentation>
 					<xs:documentation>Token of a storage configuration.</xs:documentation>
 				</xs:annotation>
 			</xs:element>
+			<xs:element name="StorageStrategy" type="xs:string" minOccurs="0">
```
is this parameter is applicable only for the camera which supports cloud object storage? I see StorageStrategy->Local also should support export recorded data to the external storage using ExportRecordedSegments.

jflevesque-genetec replied:
TargetConfiguration was added specifically for cloud recording, so no, it would not affect anything related to Profile G and local recording

jflevesque-genetec committed:
Add missing param in the specification

jflevesque-genetec committed:
Rename OmitSeqNumber to ForceZeroCounter

jflevesque-genetec committed:
Fix a typo

jflevesque-genetec committed:
Fix namespaces from move from search service

jflevesque-genetec committed:
Minor updates based on San Diego F2F and prototyping

kieran242 commented on doc/RecordingControl.xml:
```diff
@@ -2285,6 +2409,108 @@ secfrac = "." 1*6DIGIT</programlisting>
+        Segment stability is essential for this feature: each frame must be part of exactly one segment and each segment must always be immutable in time.
+        Additionnally, each segment must start with a keyframe to make it independent from the previous one. Some devices may record frames in a proprietary 
+        format, so the segments won't exist until exported to the cloud storage. Since the device must be able to list them beforehand, we suggest a simple
+        algorithm to determin how many segments will be in each time range and which frames will be part of them.
```
```suggestion
        algorithm to determine how many segments will be in each time range and which frames will be part of them.
```

kieran242 commented on doc/RecordingControl.xml:
```diff
@@ -2285,6 +2409,108 @@ secfrac = "." 1*6DIGIT</programlisting>
+        algorithm to determin how many segments will be in each time range and which frames will be part of them.
+      </para>
+      <para>
+        The figure <xref linkend="_refSegmentExportStabilityExample" /> illustrate how a global real time clock timestamp (that doesn't reset on device reboot 
```
```suggestion
        The figure <xref linkend="_refSegmentExportStabilityExample" /> illustrates how a global real time clock timestamp (that doesn't reset on device reboot 
```

kieran242 commented on doc/RecordingControl.xml:
```diff
@@ -2285,6 +2409,108 @@ secfrac = "." 1*6DIGIT</programlisting>
+      </para>
+      <para>
+        The figure <xref linkend="_refSegmentExportStabilityExample" /> illustrate how a global real time clock timestamp (that doesn't reset on device reboot 
+        or resync on NTP) could be segmented in time ranges for the target segment duration with a modulus operation. Then each boundary of these ranges is 
```
```suggestion
        or resync on NTP) could be segmented in time ranges for the target segment duration with a modulus operation. Then each boundary of these ranges are 
```

jflevesque-genetec committed:
Update doc/RecordingControl.xml

Co-authored-by: Kieran McCartan <70752937+kieran242@users.noreply.github.com>

jflevesque-genetec committed:
Update doc/RecordingControl.xml

Co-authored-by: Kieran McCartan <70752937+kieran242@users.noreply.github.com>

jflevesque-genetec committed:
Update doc/RecordingControl.xml

Co-authored-by: Kieran McCartan <70752937+kieran242@users.noreply.github.com>

HansBusch commented on wsdl/ver10/schema/onvif.xsd:
```diff
@@ -7721,6 +7726,11 @@ and sample rate. </xs:documentation>
 					<xs:documentation>Token of a storage configuration.</xs:documentation>
 				</xs:annotation>
 			</xs:element>
+			<xs:element name="StorageStrategy" type="xs:string" minOccurs="0">
```
Please move the new element to then end immediately before the xs:any to avoid breaking backward compatibility. As alternative you may defined it as an attribute.

melanconj committed:
Merge with override. Move StorageStrategy before the ##any

HansBusch committed:
Unify maximum number of results request parameter and make it everywhere optional.

HansBusch committed:
Segment Playback XML syntax fixes (#540)

* Fix xref syntax error.

* Add extension points and soap binding.

* Fix syntactical errors resulting in wrong or missing stubs with Visual Studio:

HansBusch committed:
Fix StopExportRecordedSegments response. VisualStudio needs empty sequence to build stubs.

HansBusch committed:
Combine encapsulate response parameters. (#550)

Unify boolean name HasMoreResults.

HansBusch commented on doc/RecordingControl.xml:
```diff
@@ -1673,6 +1673,130 @@ Change Request 2061, 2063, 2065, 2109</revremark>
+            <para role="param">RecordingToken [tt:RecordingReference]</para>
+            <para role="param">StorageToken - optional [tt:ReferenceToken]</para>
+            <para role="text">If provided, overrides the Target's storage configuration.</para>
+            <para role="param">Track - optional [tt:TrackType]</para>
```
Found another anoying issue as Visual Studio doesn't support optional enums parameters in the top level. Any enum becomes mandatory. As we don't have an All member and the enumeration can't be extended as it is hard wired in multiple existing methods I suggest to change the parameter to xs:string bound to tt:TrackType.

jmelancongen replied:
Ah, that's annoying... Let's go with that I guess then

HansBusch committed:
Use xs:string for passing tt:TrackType to workaround Visual Studio optional parameter limitation.

HansBusch commented on wsdl/ver10/recording.wsdl:
```diff
@@ -646,6 +644,212 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+							</xs:annotation>
+						</xs:element>
+						<xs:element name="StartTime" type="xs:dateTime">
+							<xs:annotation>
```
Shouldn't StartTime be optional? In the most simplistic case the client wants to get information about all segments.

HansBusch committed:
Improve export specification requirements (#541)

* Add capability for segment export.
* Detail requirements on segment export functions.
* Add faults.
* Clarify getting export state information shortly after completion.

kieran242 commented:
@jflevesque-genetec @HansBusch I re-reviewed this PR and #541 is there much more to do on this PR for 25.06 release. 

There is just a need for another approval correct?

HansBusch committed:
Document storage strategy and add configuration faults in case of mis… (#559)

* Document storage strategy and add configuration faults in case of missing or malfunction of local storage.

* Use standard fault code.

HansBusch committed:
Remove requirement for StoragePath.
Improve readability on handling multiple export requests.
Require chronological upload order to ensure incrementing counter.

HansBusch commented:
Accidentally directly pushed latest update instead of creating a PR. Removal of StoragePath aligned with Jose, but editorial change to improve readability needs review.

Note that I didn't remove the statement`, unless the OmitSequenceNumber is set to True.`. As the parameter isn't defined in schema and the path alogrithm bnf require the counter either the sentence has to be removed or the specs updated.

jflevesque-genetec committed:
Add fault section for ListRecordedSegments

jflevesque-genetec committed:
Add another fault code

jflevesque-genetec committed:
Fix typo

jflevesque-genetec committed:
Include optional alias when exporting segments

jflevesque-genetec committed:
Remove GetExportRecordedSegmentsState

jflevesque-genetec committed:
Reword when the event is being raised

jflevesque-genetec committed:
Add any element

jflevesque-genetec committed:
Fix typo

jflevesque-genetec committed:
Change Alias to SimpleItem

jflevesque-genetec committed:
Merge pull request #582 from onvif/video/segment-based-playback-remove-getstate

Remove GetExportRecordedSegmentsState

jflevesque-genetec committed:
Merge pull request #581 from onvif/video/segment-based-playback-alias

Include optional alias when exporting segments

74qb commented:
@jflevesque-genetec @HansBusch 
During prototyping I realized that specification is not saying anything about the span length.
We can assume that span size is defined by number of export jobs for specified time range or use span duration from recording configuration. Whatever will be decided it should be mentioned in the specification to avoid confusion.

jflevesque-genetec committed:
Clarify wording for ExportRecordedSegments

kieran242 reviewed (approved):
@jflevesque-genetec it was some time since I reviewed this PR and I have read all the contributions again and the full text and WSDL updates and I am happy to approve. I am aware this is in the Prototyping stage and this could change with Minor update.

Approving for now.

jflevesque-genetec committed:
Finish clean up of wsdl for stuff related to GetExportRecordedSegmentState

sujithhanwha commented on doc/RecordingControl.xml:
```diff
@@ -2364,6 +2522,109 @@ secfrac = "." 1*6DIGIT</programlisting>
+        a time-range of segments. A device shall accept multiple request but may queue and process
+        them sequentially. Segments shall be uploaded in chronological order. Progress can be
+        tracked via the AsynchronousOperationStatus event and the method
+          <literal>GetExportRecordedSegmentState</literal>. Operations can be cancelled at any time
```
@jflevesque-genetec 
document also needs to be updated, **GetExportRecordedSegmentState** is removed now.

jflevesque-genetec committed:
Update revision history

jflevesque-genetec committed:
Add Fault if StorageStrategy is not supported

jflevesque-genetec committed:
Clarified StorageStrategy usage

jflevesque-genetec committed:
Add clarification for segment inclusion based on time range

jflevesque-genetec committed:
Clean up leftover mentions for stuff that were dropped

jflevesque-genetec committed:
Replace must by shall

jflevesque-genetec committed:
Merge development into branch

melanconj committed:
Fix document

melanconj committed:
Merge branch 'development' into video/segment-based-playback

melanconj committed:
Fix badly fixed merge conflict

melanconj committed:
Merge branch 'development' into video/segment-based-playback
